### PR TITLE
Add job for extracting tables to CloudStorage

### PR DESCRIPTION
**Pull request by tseaver**
~~Uses #1064 as a base.~~

### CODE REVIEW

**Comment by tseaver** [comment]
Rebased after landing #1064.

File: gcloud/bigquery/client.py [hunk 1]
```diff
@@ -133,3 +134,22 @@ def copy_table(self, name, destination, *sources):
         :returns: a new ``CopyJob`` instance
         """
         return CopyJob(name, destination, sources, client=self)
+
+    def extract_table_to_storage(self, name, source, *destination_uris):
+        """Construct a job for copying one or more tables into another table.
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: `to_storage` or "into another table"?

File: gcloud/bigquery/client.py [hunk 2]
```diff
@@ -133,3 +134,22 @@ def copy_table(self, name, destination, *sources):
...
+
+        :type source: :class:`gcloud.bigquery.table.Table`
+        :param source: table to be extracted.
+
+        :type destination_uris: sequence of string
+        :param destination_uris: URIs of CloudStorage file(s) into which
+                                 table data is to be extracted.
+
+        :rtype: :class:`gcloud.bigquery.job.ExtractTableToStorageJob`
+        :returns: a new ``ExtractTableToStorageJob`` instance
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: This is kind of redundant :smile:

How about, "The job to be executed." or something equally redundant.

File: gcloud/bigquery/job.py [hunk 3]
```diff
@@ -939,3 +939,204 @@ def _build_resource(self):
         self._populate_config_resource(configuration)
 
         return resource
+
+
+class _ExtractConfiguration(object):
```
Comment (dhermes) [inline gcloud/bigquery/job.py]: Why is this one private / protected but the other two are public?

File: gcloud/bigquery/job.py [hunk 4]
```diff
@@ -939,3 +939,204 @@ def _build_resource(self):
...
+        :rtype: string, or ``NoneType``
+        :returns: The value as set by the user, or None (the default).
+        """
+        return self._configuration._destination_format
+
+    @destination_format.setter
+    def destination_format(self, value):
+        """Update destination_format.
+
+        :type value: boolean
```
Comment (dhermes) [inline gcloud/bigquery/job.py]: Boolean? Copy-pasta?

File: gcloud/bigquery/job.py [hunk 5]
```diff
@@ -939,3 +939,204 @@ def _build_resource(self):
...
+        :rtype: string, or ``NoneType``
+        :returns: The value as set by the user, or None (the default).
+        """
+        return self._configuration._compression
+
+    @compression.setter
+    def compression(self, value):
+        """Update compression.
+
+        :type value: boolean
```
Comment (dhermes) [inline gcloud/bigquery/job.py]: String?

File: gcloud/bigquery/job.py [hunk 6]
```diff
@@ -939,3 +939,204 @@ def _build_resource(self):
...
+            raise ValueError("Pass a boolean")
+        self._configuration._print_header = value
+
+    @print_header.deleter
+    def print_header(self):
+        """Delete print_header."""
+        del self._configuration._print_header
+
+    def _populate_config_resource(self, configuration):
+
```
Comment (dhermes) [inline gcloud/bigquery/job.py]: 3rd in a row. Am I missing something on these newlines / no docstrings?

File: gcloud/bigquery/job.py [hunk 7]
```diff
@@ -939,3 +939,204 @@ def _build_resource(self):
...
+            configuration['compression'] = self.compression
+        if self.destination_format is not None:
+            configuration['destinationFormat'] = self.destination_format
+        if self.field_delimiter is not None:
+            configuration['fieldDelimiter'] = self.field_delimiter
+        if self.print_header is not None:
+            configuration['printHeader'] = self.print_header
+
+    def _build_resource(self):
+        """Generate a resource for ``begin``."""
```
Comment (dhermes) [inline gcloud/bigquery/job.py]: This won't show up in gen'd docs so it doesn't matter but 

```
:meth:`begin`
```

is preferred.

File: gcloud/bigquery/client.py (hunk 1 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: "Into another table" is the `copy_table` one -- this one extracts the data (as CSV, newline-delmited JSON, or one other weird format) to storage blob(s).

Comment (dhermes) [inline gcloud/bigquery/client.py]: OK then the first line of the docstring needs to be updated (it's what I had in quotes).

Comment (tseaver) [inline gcloud/bigquery/client.py]: a94c14e

File: gcloud/bigquery/client.py (hunk 2 above)
Comment (tseaver) [inline gcloud/bigquery/client.py]: 35697d4

File: gcloud/bigquery/job.py (hunk 7 above)
Comment (tseaver) [inline gcloud/bigquery/job.py]: 0d3317f

File: gcloud/bigquery/job.py (hunk 6 above)
Comment (tseaver) [inline gcloud/bigquery/job.py]: 8939784

File: gcloud/bigquery/job.py (hunk 5 above)
Comment (tseaver) [inline gcloud/bigquery/job.py]: 98b7879

File: gcloud/bigquery/job.py (hunk 4 above)
Comment (tseaver) [inline gcloud/bigquery/job.py]: Copy-pasta, indeed.  98b7879

File: gcloud/bigquery/client.py [hunk 8]
```diff
@@ -95,7 +96,6 @@ def dataset(self, name):
         :param name: Name of the dataset.
 
         :rtype: :class:`gcloud.bigquery.dataset.Dataset`
-        :returns: a new ``Dataset`` instance
```
Comment (dhermes) [inline gcloud/bigquery/client.py]: Though redundant, we should view `:returns:` as "required". I didn't imagine 35697d49ea3a0ed76c565f7ecb68b4c8394ff196 would come from my tongue-in-cheek comment.

Comment (tseaver) [inline gcloud/bigquery/client.py]: Your comment pointed out that the element was redundant:  it added no new information to the docstring beyond the summary line and `:rtype:`.  I will rebase out 35697d49ea3a0ed76c565f7ecb68b4c8394ff196

**Comment by dhermes** [comment]
Are we all good here? AFAICT we got the `:returns:` issue sorted out. If yes, LGTM and feel free to merge. Ping me on the dependent PRs once rebased.

**Pull request merged**